068-access-policies-in-jwt.md

```diff
@@ -0,0 +1,251 @@
+and Keycloak authorizer.
+It also supports custom authorizers, which can be used to point to the implementation of a custom authorizer class.
+There is currently no inbuilt support for JWT-based authorization.
+If a user wants to use JWT-based authorization, they have to implement their own authorizer and use the custom authorizer feature.
+As the documentation states:
+
+> In addition to the Kafka custom resource configuration, the JAR file containing the custom authorizer class along with its dependencies must be available on the classpath of the Kafka broker.
+
+This proposal aims to add the JWT-based authorizer to Strimzi, so it can be selected in the configuration as
+another option apart from the simple and Keycloak ones, without having to implement a custom authorizer.
+
+## Motivation
+
+Having a generic solution is always preferable as it allows for less effort to be maintained.
+While we cannot achieve a perfect generic solution due to the nature of OAuth2, and how different
+identity providers are implemented, we can strive to create a solution as generic as possible.
+
+Passing additional data in the JWT seems to be supported by the majority of identity providers.
+Therefore, we can use this to pass the access policies in the JWT.
+This solution will work in most cases and with most identity providers, however it will not
+scale to infinity.
+It will, however, be decoupled from Keycloak and therefore be more generic.
+
+## Why JWT as an authorization mechanism?
+
+The official [JWT](https://jwt.io/introduction) lists some use cases:
+
+> Authorization: This is the most common scenario for using JWT. Once the user is logged in, each subsequent request will include the JWT, allowing the user to access routes, services, and resources that are permitted with that token. Single Sign On is a feature that widely uses JWT nowadays, because of its small overhead and its ability to be easily used across different domains.
+
+JWT supports custom claims, which can be used to pass additional claims/roles.
+For example [Spring Security](https://docs.spring.io/spring-authorization-server/reference/guides/how-to-custom-claims-authorities.html) can pass both claims and roles in the JWT,
+which can be used to authorize the user.
+
+All OAuth2 providers support custom claims in the JWT in one way or another.
+We have listed the most popular ones in the [Support for roles or claims in popular identity providers](#support-for-roles-or-claims-in-popular-identity-providers) section with links to the documentation
+on how custom claims or roles are supported.
+
+The concept of roles can be directly mapped to the concept of access policies in Kafka.
+
+## Proposal
+
+Identity providers support passing custom fields in the JWT token.
+Some providers allow the data type to be string, while others allow a list of strings.
+There are not many (if any) identity providers that allow a list of objects.
```

Comment:
Are we going to support regex expressions for the topic name? I think this would help to support use cases where you have many topics with a common prefix (e.g. the app name).

Reply:
That is an interesting idea. We are currently using an implementation that is a subset of glob patterns - handling wildcard matches to allow matching a prefix or a suffix. Considering that we want to compress the ACLs as much as possible, regex could be a good fit, but my concern is that it might overcomplicate it. I cannot think of another situation apart from common prefix/suffix where you'd need the power of regex.
One more concern I have with regex is that it would be hard to debug. Glob patterns are a bit more expressive and have more consistent performance impact.

Reply:
I agree on the points you made in favor of glob patterns for name matching. Can you add them to the proposal with some examples? Also, I would add regex to rejected alternatives with the motivation.

Reply:
done